### PR TITLE
Tower job template update

### PR DESCRIPTION
**Pull request by john-westcott-iv**
##### SUMMARY

Conversion of tower_job_template module off tower-cli

##### ISSUE TYPE

 - Feature Pull Request

##### COMPONENT NAME

 - Collection

##### AWX VERSION

```
awx: 9.3.0
```

##### ADDITIONAL INFORMATION

```

```

### CODE REVIEW

File: awx_collection/plugins/module_utils/tower_api.py [hunk 1]
```diff
@@ -94,6 +94,17 @@ def __init__(self, argument_spec, **kwargs):
 
         self.session = Request(cookies=CookieJar(), validate_certs=self.verify_ssl)
 
+    def extract_errors_from_response(self, response):
+        if 'json' in response:
+            if '__all__' in response['json']:
+                return response['json']['__all__'][0]
+            elif 'error' in response['json']:
+                return response['json']['error']
```
Comment (AlanCoding) [inline awx_collection/plugins/module_utils/tower_api.py]: what about "detail" and "msg", I think those should be inspected too because it's sometimes a thing.

Also, best to do this in a loop.

And go ahead with the duck typing, if it's a list, then make it into a string. I don't like `[0]`, I'd rather do `\n'.join()`. Your choice for the separator.

Comment (john-westcott-iv) [reply awx_collection/plugins/module_utils/tower_api.py]: Do you have examples of detail and msg? Are those at the root of the object returned? I'll take care of the loop and joins in the mean time.

File: awx_collection/plugins/module_utils/tower_api.py [hunk 2]
```diff
@@ -440,53 +459,113 @@ def delete_if_needed(self, existing_item, handle_response=True, on_delete=None):
...
+                # First get the existing data from the endpoint
+                existing_data = self.get_endpoint(endpoint)
+
+                # If the endpont data is identical, just continue on to the next item
+                if existing_data['json'] == post['data']:
+                    continue
+
+                with open('/tmp/john', 'w') as f:
+                    import json
+                    f.write(json.dumps(post['data']))
```
Comment (AlanCoding) [inline awx_collection/plugins/module_utils/tower_api.py]: Not going to lie, when I'm in these situations I `raise Exception(data)` and run the test

File: awx_collection/plugins/module_utils/tower_api.py [hunk 3]
```diff
@@ -440,53 +459,113 @@ def delete_if_needed(self, existing_item, handle_response=True, on_delete=None):
...
+
+                # If this item no longer has data, lets remove it
+                if post['data'] in ('', [], {}):
```
Comment (AlanCoding) [inline awx_collection/plugins/module_utils/tower_api.py:542]: This could cause weird changed=True cases, since `existing_data['json'] == post['data']` will be unaware of cases which are different in syntax, but both indicate various version of nothing.

File: awx_collection/plugins/module_utils/tower_api.py [hunk 4]
```diff
@@ -440,53 +459,113 @@ def delete_if_needed(self, existing_item, handle_response=True, on_delete=None):
...
+                    response = self.delete_endpoint(endpoint)
+                    # Tower job template requires a 200
+                    # Left as an array as for future additional delete returns
+                    if response['status_code'] in [200]:
+                        self.json_output['deleted_additional_post'] = True
```
Comment (AlanCoding) [inline awx_collection/plugins/module_utils/tower_api.py]: hmm... maybe it would be good to report _which_ endpoint was deleted, considering you wrote this to handle multiples

File: awx_collection/plugins/module_utils/tower_api.py [hunk 5]
```diff
@@ -440,53 +459,113 @@ def delete_if_needed(self, existing_item, handle_response=True, on_delete=None):
...
+                self.fail_json(msg="Addition post type of {0} is not valid for {1}".format(association_type, item_type))
+
+    def process_associations(self, item, item_type, associations):
+        # Process any associations with this item
+        if associations is not None:
+            for association_type in associations:
+                if item.get('related', {}).get(association_type, None):
+                    self.modify_associations(item['related'][association_type], associations[association_type])
+                else:
+                    self.fail_json(msg="Association type of {0} is not valid for {1}".format(association_type, item_type))
```
Comment (AlanCoding) [inline awx_collection/plugins/module_utils/tower_api.py:576]: this stuff should get cleaned up after rebasing

File: awx_collection/plugins/modules/tower_job_template.py [hunk 6]
```diff
@@ -12,409 +13,532 @@
...
+    elif state == 'present':
+        # If the state was present and we can let the module build or update the existing item, this will return on its own
+        module.create_or_update_if_needed(
+            existing_item, new_fields,
+            endpoint='job_templates', item_type='job_template',
+            associations={
+                'credentials': credentials_ids,
+                'labels': labels_ids,
+            },
+            additional_posts=additional_posts,
```
Comment (AlanCoding) [inline awx_collection/plugins/modules/tower_job_template.py:578]: To get a sense of how this compares to the current state, I wrote this, partially influenced by your code:

https://github.com/ansible/awx/pull/6291/commits/581d495f66ee831ffbf4f61422664a5e139c8886

I'm trying to grasp how your PR winds up with much more code, and it may be largely in the error handling.

File: awx_collection/plugins/module_utils/tower_api.py (hunk 2 above)
Comment (john-westcott-iv) [reply awx_collection/plugins/module_utils/tower_api.py]: Yea, I should probably do something better but exceptions in Ansible can be a tad ugly to go through. Especially just when I want to try and dump a response data out. Removed the debug.

**Comment by AlanCoding** [comment]
I don't know if this is going to work for approval templates. These are created in a related POST from the WFJT node. This should be similar to what you're doing here, but the details go a little crazy.

Looking at what the UI does, it makes a POST:

url `/api/v2/workflow_job_template_nodes/10/create_approval_template/`

```json
{
  "name": "fffff",
  "description": null,
  "timeout": 0
}
```

In response, the API gives a 201 and a full serialization of the related object. Like:

```json
{
  "id": 17,
  "type": "workflow_approval_template",
  "url": "/api/v2/workflow_approval_templates/17/",
  "related": {
    "created_by": "/api/v2/users/1/",
    "modified_by": "/api/v2/users/1/",
    "jobs": "/api/v2/workflow_approval_templates/17/approvals/"
  },
  "summary_fields": {
    "workflow_job_template": {
      "id": 16,
      "name": "has approval node",
      "description": ""
    },
    "created_by": {
      "id": 1,
      "username": "admin",
      "first_name": "",
      "last_name": ""
    },
    "modified_by": {
      "id": 1,
      "username": "admin",
      "first_name": "",
      "last_name": ""
    }
  },
  "created": "2020-03-20T15:18:28.757683Z",
  "modified": "2020-03-20T15:18:28.757696Z",
  "name": "fffff",
  "description": "",
  "last_job_run": null,
  "last_job_failed": false,
  "next_job_run": null,
  "status": "never updated",
  "timeout": 0
}
```

@beeankha do you have any thoughts on how to handle approval nodes?

File: awx_collection/plugins/module_utils/tower_api.py (hunk 5 above)
Comment (john-westcott-iv) [reply awx_collection/plugins/module_utils/tower_api.py:576]: I feel like we have a little bit of a mess on our hands with all of the things flying around @beeankha and I are going to try and cleanup and rebase on Monday.

File: awx_collection/plugins/module_utils/tower_api.py (hunk 3 above)
Comment (john-westcott-iv) [reply awx_collection/plugins/module_utils/tower_api.py:542]: Do you have examples of that? This was a first crack at defining some kind of pattern, if we can't make it work we can try something else.

File: awx_collection/plugins/module_utils/tower_api.py (hunk 4 above)
Comment (john-westcott-iv) [reply awx_collection/plugins/module_utils/tower_api.py]: With the next push it will now spit out something like:
```
        "added_additional_post": "/api/v2/job_templates/112/survey_spec/",
        "added_an_association": [
            "/api/v2/job_templates/112/credentials/[1]",
            "/api/v2/job_templates/112/credentials/[14]",
            "/api/v2/job_templates/112/labels/[120]",
            "/api/v2/job_templates/112/labels/[121]"
        ],
```
I added the [<id>] to the "endpoint" so you would know better what was going on rather than just getting a bunch of meaningless /labels/ endpoints.

File: awx_collection/plugins/module_utils/tower_api.py [hunk 7]
```diff
@@ -518,27 +587,21 @@ def create_if_needed(self, existing_item, new_item, endpoint, on_create=None, it
...
-                    self.fail_json(msg="Unable to create {0} {1}: {2}".format(item_type, item_name, response['json']))
-                else:
-                    self.fail_json(msg="Unable to create {0} {1}: {2}".format(item_type, item_name, response['status_code']))
+                self.fail_json(msg="Unable to create {0} {1}: {2}".format(item_type, item_name, self.extract_errors_from_response(response)), **{'response': response})
 
-        # Process any associations with this item
-        if associations is not None:
-            for association_type in associations:
-                self.modify_associations(response, associations[association_type])
+        self.process_associations(response['json'], item_type, associations)
```
Comment (AlanCoding) [inline awx_collection/plugins/module_utils/tower_api.py]: This still isn't going to work if `existing_item` is non-null when the method is called.

https://github.com/ansible/awx/pull/6369

I'm sorry this is going to mess up the diff management even more. I really want us to focus on having these stable and testable as more features are added. If you want to put up a minor patch in a way that better aligns with with your direction, that's fine, but I want to address the minor patch first, before pulling large wide-ranging changes to the common code.

File: awx_collection/plugins/modules/tower_job_template.py [hunk 8]
```diff
@@ -12,409 +13,532 @@
...
       description:
         - The job type to use for the job template.
       required: False
-      choices: ["run", "check"]
       type: str
+      default: 'run'
+      choices:
+        - 'run'
+        - 'check'
+        - 'scan'
```
Comment (AlanCoding) [inline awx_collection/plugins/modules/tower_job_template.py:51]: Previously, we explicitly got rid of the "scan" option. I thought the API had purged it, but it is also in the AWX collection list of changes.

**Comment by beeankha** [comment]
recheck

**Comment by beeankha** [comment]
recheck

**Comment by beeankha** [comment]
recheck

File: awx_collection/test/awx/test_team.py [hunk 9]
```diff
@@ -47,7 +48,10 @@ def test_modify_team(run_module, admin_user):
     }, admin_user)
     team.refresh_from_db()
     result.pop('invocation')
+    result.pop('field_changes')
     assert result == {
+        "needed_update": True,
+        "updated": False,
```
Comment (AlanCoding) [inline awx_collection/test/awx/test_team.py]: This isn't something consistent enough that we could document in the `RETURNS` string. If it won't buy us that, I don't see why we should do it.

If it's created, then it returns the `created` key, and if it's updated it returns the `updated` key. But neither of these keys are present if it's not created or updated. The proposal in https://github.com/ansible/awx/issues/6367 is that other status like `approved` would also be present. Then should `started` and `finished` be status types too?

I'd advocate that returning `state` might be best, even if it's a repeat of the input parameter. That's documentable, and not always obvious because it has a default.

Comment (john-westcott-iv) [reply awx_collection/test/awx/test_team.py]: I added some of this as almost "Tracking" so that you could look at the returned data and know what it did. I thought that might also help us prepare for diff mode. i.e. instead of just `changed: True` why was it actually changed. This has also helped in development of Tower credential to know why a credential is being reported as changed.

I think we should be able to set them in the returned field because they can have an optional description on them (I believe). i.e.
```
needed_update:
    description: If the item needed to be updated
    returned: when item already exists
    type: bool
    sample: True
```

Comment (AlanCoding) [reply awx_collection/test/awx/test_team.py]: If there's going to be churn on this, we might be better off to rewrite the (unit) test assertions so that they just don't assert on the full data - and develop something after things have stabilized. For the purposes of testing / tracking specific behaviors, that makes sense to me, I just worry if the key naming seems too general. But I also can't connect the dots from these changes to the JT module conversion. A separate PR might be nice.

File: awx_collection/plugins/modules/tower_group.py [hunk 10]
```diff
@@ -112,9 +112,21 @@ def main():
...
+    hosts_ids = None
+    if hosts is not None:
+        hosts_ids = []
+        for item in hosts:
+            hosts_ids.append(module.resolve_name_to_id('hosts', item))
+    children_ids = None
+    if children is not None:
+        children_ids = []
+        for item in children:
+            children_ids.append(module.resolve_name_to_id('groups', item))
```
Comment (AlanCoding) [inline awx_collection/plugins/modules/tower_group.py]: this directly conflicts with `association_fields`, I think this would be better split out into another PR that just changes the field name, which I agree should be done.

Comment (john-westcott-iv) [reply awx_collection/plugins/modules/tower_group.py]: Agreed, this PR should have just been for tower_job_template. @beeankha let me know if you want to pull this out of here or if I should?

File: awx_collection/test/awx/test_team.py (hunk 9 above)
Comment (john-westcott-iv) [reply awx_collection/test/awx/test_team.py]: The only reason they were in here was because I was hitting a condition where the JT was always returning changed = True. Instead of doing my old: `with open('/tmp/john', 'w') as f ....` I tried to get smart and add these additional fields into the module_util so that I could tell what it was trying to do internally. Incidentally, I needed these things when working on tower_credential to track down the issue with the inputs hash.

File: awx_collection/plugins/modules/tower_group.py (hunk 10 above)
Comment (beeankha) [reply awx_collection/plugins/modules/tower_group.py]: I can do this, and put the changes into a new `tower_group.py`-specific PR.

File: awx_collection/plugins/modules/tower_job_template.py [hunk 11]
```diff
@@ -12,409 +13,562 @@
...
     argument_spec = dict(
-        name=dict(required=True),
-        description=dict(default=''),
-        job_type=dict(choices=['run', 'check']),
-        inventory=dict(default=''),
-        project=dict(required=True),
-        playbook=dict(required=True),
+        name=dict(required=True, type='str'),
+        new_name=dict(required=False, type='str'),
+        description=dict(required=False, type='str', default=''),
```
Comment (squidboylan) [inline awx_collection/plugins/modules/tower_job_template.py]: I'm moderately concerned about all the defaults here, mostly in the case of modifying a job_template, if we have a default and the user doesn't specify a value for that parameter we'll overwrite that parameter with the default, we do need defaults for a lot of these things, but only for new objects. I think we ran into this issue with a different module, i forget which one, maybe user

Comment (AlanCoding) [reply awx_collection/plugins/modules/tower_job_template.py]: we've discussed this in some other places too. I'm in agreement, my stance is that `state` should have a default, and the rest can go.

File: awx_collection/plugins/module_utils/tower_api.py (hunk 5 above)
Comment (AlanCoding) [reply awx_collection/plugins/module_utils/tower_api.py:576]: The existing item, `item` in this case will be `None` in some cases. This comes from the case where the object is created, and has associations, and the fix I applied to that doesn't look like it will work here. Some test coverage for this was introduced in https://github.com/ansible/awx/pull/6369. What changed with the processing of new items in this PR, and how is it expected that the case of create-and-associate will work with the call patterns as they were changed here?

File: awx_collection/plugins/module_utils/tower_api.py [hunk 12]
```diff
@@ -551,12 +635,12 @@ def create_if_needed(self, existing_item, new_item, endpoint, on_create=None, it
                 self.modify_associations(sub_endpoint, associations[association_type])
 
         # If we have an on_create method and we actually changed something we can call on_create
-        if on_create is not None and self.json_output['changed']:
+        if on_create is not None and self.json_output['created']:
```
Comment (AlanCoding) [inline awx_collection/plugins/module_utils/tower_api.py]: If you modify a project's URL or something like that, you still need to wait for update because it will auto-update in those cases.

**Comment by AlanCoding** [comment]
```python
@pytest.mark.django_db
def test_job_template_with_new_credentials(
        run_module, admin_user, project, inventory,
        machine_credential, vault_credential):
    result = run_module('tower_job_template', dict(
        name='foo',
        playbook='helloworld.yml',
        project=project.name,
        inventory=inventory.name,
        credentials=[machine_credential.name, vault_credential.name]
    ), admin_user)
    assert not result.get('failed', False), result.get('msg', result)
    assert result.get('changed', False), result
    jt = JobTemplate.objects.get(pk=result['id'])

    assert set([machine_credential.id, vault_credential.id]) == set([
        cred.pk for cred in jt.credentials.all()])

    prior_ct = ActivityStream.objects.count()
    result = run_module('tower_job_template', dict(
        name='foo',
        playbook='helloworld.yml',
        project=project.name,
        inventory=inventory.name,
        credentials=[machine_credential.name, vault_credential.name]
    ), admin_user)
    assert not result.get('failed', False), result.get('msg', result)
    assert not result.get('changed', True), result
    jt.refresh_from_db()
    assert result['id'] == jt.id

    assert set([machine_credential.id, vault_credential.id]) == set([
        cred.pk for cred in jt.credentials.all()])
    assert ActivityStream.objects.count() == prior_ct


@pytest.mark.django_db
def test_job_template_with_survey_spec(run_module, admin_user, project, inventory, survey_spec):
    result = run_module('tower_job_template', dict(
        name='foo',
        playbook='helloworld.yml',
        project=project.name,
        inventory=inventory.name,
        survey_spec=survey_spec,
        survey_enabled=True
    ), admin_user)
    assert not result.get('failed', False), result.get('msg', result)
    assert result.get('changed', False), result
    jt = JobTemplate.objects.get(pk=result['id'])

    assert jt.survey_spec == survey_spec

    prior_ct = ActivityStream.objects.count()
    result = run_module('tower_job_template', dict(
        name='foo',
        playbook='helloworld.yml',
        project=project.name,
        inventory=inventory.name,
        survey_spec=survey_spec,
        survey_enabled=True
    ), admin_user)
    assert not result.get('failed', False), result.get('msg', result)
    assert not result.get('changed', True), result
    jt.refresh_from_db()
    assert result['id'] == jt.id

    assert jt.survey_spec == survey_spec
    assert ActivityStream.objects.count() == prior_ct
```

I'd suggest adding these, replacing the existing credentials test.

I also believe that both tests should be expanded to modify the credentials and the survey spec. The credential modification scenarios are a lot, so I'd pack them all into 1 test (add, remove, stay the same). You could piggyback on these cases to these tests, or add new tests.

Right now I get:

```
awx_collection/test/awx/test_job_template.py ....F                                                                                                                            [100%]

===================================================================================== FAILURES ======================================================================================
________________________________________________________________________ test_job_template_with_survey_spec _________________________________________________________________________
Traceback (most recent call last):
  File "/Users/alancoding/Documents/tower/awx_collection/test/awx/test_job_template.py", line 149, in test_job_template_with_survey_spec
    assert jt.survey_spec == survey_spec
AssertionError: assert {} == {'description... False, ...}]}
  Right contains 3 more items:
  {'description': 'test',
   'name': 'test',
   'spec': [{'default': 'mydef',
             'index': 0,
             'question_name': 'my question?',
             'required': False,...
  
  ...Full output truncated (4 lines hidden), use '-vv' to show
============================================================================ 1 failed, 4 passed in 4.13s ============================================================================
```

File: awx_collection/test/awx/test_inventory_source.py [hunk 13]
```diff
@@ -42,6 +42,7 @@ def test_inventory_source_create(run_module, admin_user, base_inventory, project
     assert result == {
         'id': inv_src.id,
         'name': 'foo',
+        'created': True,
```
Comment (AlanCoding) [inline awx_collection/test/awx/test_inventory_source.py]: just change the whole thing to `assert result['id'] == inv_src.id` because it is being said here that the others are not supported.

The name return value might be alright, but it's not given in all cases. No one has any idea if should be there or not.

File: awx_collection/tools/vars/associations.yml [hunk 14]
```diff
@@ -0,0 +1,16 @@
...
+  job_templates:
+    - related_item: credentials
+      endpoint: credentials
+      description: "The credentials used by this job template"
+      required: false
+  groups:
+    - related_item: hosts
+      endpoint: hosts
+      description: "The hosts associated with this group"
+      required: false
```
Comment (AlanCoding) [inline awx_collection/tools/vars/associations.yml]: I think `required` shouldn't be a thing here, and the arg spec should just leave it out.

File: awx_collection/tools/vars/resolution.yml [hunk 15]
```diff
@@ -0,0 +1,6 @@
+---
+name_to_id_endpoint_resolution:
+  webhook_credential: credentials
+  project: projects
+  inventory: inventories
+  organization: organizations
```
Comment (AlanCoding) [inline awx_collection/tools/vars/resolution.yml]: I don't think we should continue to rely on name_to_id, and encoding the logic in the generator is going in the wrong directions. I would suggest that `param_to_endpoint` can just be called, and there is no need for an entire pipeline for producing this call pattern.

File: awx_collection/plugins/modules/tower_job_template.py [hunk 16]
```diff
@@ -12,409 +13,562 @@
...
+    if allow_simultaneous is not None:
+        new_fields['allow_simultaneous'] = allow_simultaneous
+    if custom_virtualenv is not None:
+        new_fields['custom_virtualenv'] = custom_virtualenv
+    if job_slice_count is not None:
+        new_fields['job_slice_count'] = job_slice_count
+    if webhook_service is not None:
+        new_fields['webhook_service'] = webhook_service
+    if webhook_credential is not None:
+        new_fields['webhook_credential'] = webhook_credential_id
```
Comment (AlanCoding) [inline awx_collection/plugins/modules/tower_job_template.py:564]: This is a lot of additional code. The old module did an outright `params.copy()`, so there was no enumeration of the field names at all, and with this PR we're losing that, and we're not getting anything for it. As a _tiny_ improvement I've listed the params in other modules to keep it slightly more brief. In this case I think the ideal solution would be to remove params that shouldn't be treated in the standard way.

File: awx_collection/plugins/modules/tower_job_template.py [hunk 17]
```diff
@@ -12,409 +13,562 @@
...
+        - Name of the vault credential to use for the job template.
+        - Deprecated, use 'credentials'.
+      version_added: 2.7
+      type: str
+    tower_oauthtoken:
+      description:
+        - The Tower OAuth token to use.
+      required: False
+      type: str
+      version_added: "3.7"
```
Comment (AlanCoding) [inline awx_collection/plugins/modules/tower_job_template.py:327]: Is there any way to find out what in the parameter spec changed?

File: awx_collection/plugins/modules/tower_job_template.py [hunk 18]
```diff
@@ -12,409 +13,562 @@
...
-    job_type: "run"
-    inventory: "Local"
-    project: "Demo"
-    playbook: "ping.yml"
-    credential: "Local"
-    state: "present"
-    tower_config_file: "~/tower_cli.cfg"
-    survey_enabled: yes
-    survey_spec: "{{ lookup('file', 'my_survey.json') }}"
-    custom_virtualenv: "/var/lib/awx/venv/custom-venv/"
```
Comment (AlanCoding) [inline awx_collection/plugins/modules/tower_job_template.py:249]: the module does need an example

File: awx_collection/plugins/modules/tower_job_template.py [hunk 19]
```diff
@@ -12,409 +13,562 @@
...
+        for item in credentials:
+            credentials_ids.append(module.resolve_name_to_id('credentials', item))
+    labels_ids = None
+    if labels is not None:
+        labels_ids = []
+        for item in labels:
+            labels_ids.append(module.resolve_name_to_id('labels', item))
+
+    # Pull in additional posts
+    additional_posts = []
```
Comment (AlanCoding) [inline awx_collection/plugins/modules/tower_job_template.py:478]: Since they're both hitting some kind of related items, it's really mismatched that the additional posts is a list while the additional associations is a dict. It's also really hard to keep up with an implied nested schema.

File: awx_collection/plugins/module_utils/tower_api.py [hunk 20]
```diff
@@ -551,12 +635,12 @@ def create_if_needed(self, existing_item, new_item, endpoint, on_create=None, it
                 self.modify_associations(sub_endpoint, associations[association_type])
```
Comment (AlanCoding) [inline awx_collection/plugins/module_utils/tower_api.py:636]: You're adding a helper method to tease out "detail" or "__all__" or other stuff from the response, and to the extent that you still want to do that, branching like what we have here in `create_if_needed` should be converted to use that helper method.

Comment (AlanCoding) [reply awx_collection/plugins/module_utils/tower_api.py:636]: Also, the method `create_if_needed` calls `modify_associations` while the other one calls `process_associations`. If it's going to be called in exactly 1 place, there's no reason to put it into a separate method.

File: awx_collection/plugins/modules/tower_job_template.py (hunk 18 above)
Comment (beeankha) [reply awx_collection/plugins/modules/tower_job_template.py:249]: Oops, will add one back in...

File: awx_collection/plugins/modules/tower_group.py (hunk 10 above)
Comment (AlanCoding) [reply awx_collection/plugins/modules/tower_group.py]: resolved separately

File: awx_collection/plugins/module_utils/tower_api.py (hunk 3 above)
Comment (AlanCoding) [reply awx_collection/plugins/module_utils/tower_api.py:542]: So I guess I'm not following this at all. This `if` should be written as `if not post['data']` unless I'm missing something. If `if 'data' in post and (not post['data'])`. Unless the intent is that `None` should not result in a DELETE. This is why the contract gets really hard to manage. There's a lot of code here for survey spec handling, I don't think it generalizes well, it doesn't align with WFJT surveys, and I would go back to the drawing board for everything with the spec.

Comment (john-westcott-iv) [reply awx_collection/plugins/module_utils/tower_api.py:542]: Correct, None would represent that the survey_spec (or any other field) was just not provided to the module. So we should not delete the survey on a None condition. The module itself is currently catching survey_spec (or any other field) being None and choosing to not add the data to the additional post parameter. However we could also go the route of just passing None back into this method (especially if we are doing auto-generation).  So instead of blocks like this:
```
   additional_posts = []
    if survey_spec is not None:
        additional_posts.append({
            'endpoint_reference': 'survey_spec',
            'data': survey_spec,
        })
```
Followed with:
```
        module.create_or_update_if_needed(
            existing_item, new_fields,
            endpoint='job_templates', item_type='job_template',
            associations={
                'credentials': credentials_ids,
                'labels': labels_ids,
            },
            additional_posts=additional_posts,
        )
```
We could just do something like:
```
        module.create_or_update_if_needed(
            existing_item, new_fields,
            endpoint='job_templates', item_type='job_template',
            associations={
                'credentials': credentials_ids,
                'labels': labels_ids,
            },
            additional_posts=[
              {'endpoint_reference': 'survey_spec','data': survey_spec,}
            ],
        )
```

It looks like survey_spec is the same for WJT as it is for JT (unless of course I'm missing something) its even posted through the same survey_spec end point. I did a quick test converting WFJT module to use additional_posts and it seemed to work (granted it wasn't a very large survey).

**Comment by beeankha** [comment]
Closed due to https://github.com/ansible/awx/pull/6491 being merged.
